week2/1376_time_inform_employees.py

```diff
@@ -0,0 +1,44 @@
+# https://leetcode.com/problems/time-needed-to-inform-all-employees/
+from collections import deque, defaultdict
+from typing import List
+
+
+class Solution:
+    def num_of_min_timeout(self, n: int, head_id: int, manager: List[int], inform_time: List[int]) -> int:
+        # BFS approach
+        queue = deque([head_id])
+        while queue:
+            current_employee = queue.popleft()
+            time_employee = inform_time[current_employee]
+            subordinates = [i for i in manager if manager[i] == current_employee]
+            if not subordinates:
+                continue
+            for s in subordinates:
+                inform_time[s] += time_employee
+                queue.append(s)
+
+        return max(inform_time)
+
+    # the function above exceeds the given time probably because the list of subordinate loops everytime
+    # for each current_employee and max() iterates through the entire list of inform_time. We can improve this
+    # by creating subordinate lists before traversing the tree and storing maximum at every level of the tree
+    def num_of_min_bfs(self, n: int, head_id: int, manager: List[int], inform_time: List[int]) -> int:
```

Comment:
Nicely written. The structure provided was not good at all for traversing. Maybe for next time, it would be nice if you can think whether using the given parameter is good to use immediately or you need to transform in a more efficient form. Because during coding interviews, I think you rarely get data structures as a graph that is immediately good for traverse. You first need to transform it.